Allows passing the list of supported features when building a supergraph

pcmanus opened the pull request:
When we parse/build a supergraph, one validation run checks that any feature in the supergraph marked for `EXECUTION` or `SECURITY` is actually supported. But the list of what was supported was currently completely hard-coded, and that means for instance that external use of the query planner (like in the Apollo router typically) could not modify this list of supported features. This commit allows that list to be provided (and default to the hard-coded one if not).

Now, passing that list down to the validation that use it was not very convenient in the existing code. Taking the query planner as an example, it was taking the supergraph as argument as a `Schema` but the validation was then done somewhere within the code of the planner. Passing the list of supported options, so it can then be passed down to the relevant code, was technically possible, but it would have been a bit ugly, because 1) the query planner is one example, but supergraphs could be use elsewhere and 2) that list of supported feature isn't really a property of the planner (which mostly doesn't care), so passing it as a planner option is not too clean.

So most of this commit is really a small refactor to introduce a `Supergraph` class, to essentially warp a schema that is known to be a supergraph schema. Places that expect a supergraph as argument now use `Supergraph` instead of just `Schema`. I'd argue that this refactor make sense regardless of the final goal of this issue for at least 2 reasons:
1. this provided nice symmetry with the `Subgraph` class we have had for a long time for very similar reasons.
2. it's a better use of the type system. Now that the query planner take a `Supergraph` instance, it's extra clear what the argument is, and if multiple methods take that instance, they know for a fact that the underlying schema is a valid supergraph and they don't have to defensively check it again.

pcmanus commented:
Note that this PR is currently on top of #2655 as those all touch code around subgraph extraction and I don't particularly care about spending my time doing rebases.

trevor-scheer commented on internals-js/src/supergraphs.ts:
```diff
@@ -21,24 +22,12 @@ const SUPPORTED_FEATURES = new Set([
 
 const coreVersionZeroDotOneUrl = FeatureUrl.parse('https://specs.apollo.dev/core/v0.1');
 
-export function buildSupergraphSchema(supergraphSdl: string | DocumentNode): [Schema, {name: string, url: string}[]] {
```
Removing this is a breaking change

trevor-scheer reviewed (commented):
I haven't bothered to try to figure out the diff in `extractSubgraphsFromSupergraph` since this is based on the other branch and there's a lot of work in there, but the rest looks good to me.

Only concern is that we've removed some of the public API, so I'll wait to hear your thoughts on that one.

pcmanus commented on internals-js/src/index.ts:
```diff
@@ -13,7 +13,6 @@ export * from './tagSpec';
 export * from './inaccessibleSpec';
 export * from './federationSpec';
 export * from './supergraphs';
-export * from './extractSubgraphsFromSupergraph';
```
> Looks like nestjs-lib is a consumer unfortunately

Quickly looking at it:
1. while the name of this package could make it sound like this is part of https://nestjs.com/, pretty sure it's not. There is zero source information on where this package is coming from and this has no README whatsoever, but a quick scan shows code comment in chinese characters, which I don't think an official component of NestJS would have. The package has also very little weekly downloads. Doesn't by itself mean that we shouldn't care about breaking it, but this is also not a `@nestjs` package with millions of weekly downloads. Besides ...
2. while this package `package.json` does list `@apollo/federation-internals` (and composition, and the query planner), a quick search makes it clear that none of those are actually used in the code (none of the `@apollo` dependencies is even imported by any file, nor is `buildSupergraphSchema` or `extract*` turn anything). So we're not breaking that one package.

pcmanus commented on internals-js/src/supergraphs.ts:
```diff
@@ -89,3 +78,52 @@ export function validateSupergraph(supergraph: Schema): [CoreFeatures, JoinSpecD
+    this.containedSubgraphs = extractSubgraphsNamesAndUrlsFromSupergraph(schema);
+  }
+
+  static build(supergraphSdl: string | DocumentNode, options?: { supportedFeatures?: Set<string> }) {
```
This is admittedly a bit of a personal preference territory, but I prefer the separately named method to emphasis that there is additional work being done (the `buildSchema` part).

trevor-scheer replied:
No worries, fine as is

pcmanus commented:
> Only concern is that we've removed some of the public API, so I'll wait to hear your thoughts on that one.

We've called that package `federation-internals` exactly because we wanted to make it clear that didn't wanted to provide backward compatibility guarantees for the API in this package. And we've broken API in this package a fair number of times, so this isn't new grounds either. Me breaking those API was not an oversight in particular. Personally, I'd stick to that but I'll add the following for context:
1. I did initially opened this against `main` out of habit/laziness. I actually do intent for this to be merged against `next` (and I've now rebased it against that branch). I mean, this is not a bug fix, so this belong to `next` and I don't really thing we have cause to do otherwise anyway. I stand by the fact that it's not worth maintaining backward compat for API of our internal packages, but I can agree that it's at least worth avoiding such change in patch releases if we can (though personally, I'd have no qualm doing so if this help fix an important bug, but it's a different discussion).
2. we kind of don't follow semver for everything in federation.  That is, we do want backward compat in minors/patches "for the fed spec and general observable behaviour", but as said before, we've broken API in some of our packages in minor release many many times since fed2.
3. I'm not saying it's impossible that anyone might actually use the API removed here in some small tooling (I have a custom tool I'm using that does), but I highly doubt this is common, and worth case, when those tool upgrade, they won't compile again and will have to make a handful of very trivial changes.

trevor-scheer reviewed (approved):
@pcmanus if it's status quo, then no worries. My only real concern was your 3rd point more out of principle than practice (since I also don't expect anyone to be using this) since it is still a public package. I think it's fine to change internal API surface in a patch / doesn't need to be minor, but `next` is good. Agree the naming is a sufficient indication that it isn't "public API".

smyrick commented:
Is it worth publishing a README for that package indicating this is considered an internal package and does not have the same semver guarantees and users should look to the main @apollo/federation package
NPM: https://www.npmjs.com/package/@apollo/federation-internals
Github: https://github.com/apollographql/federation/tree/main/internals-js

trevor-scheer commented:
Just noting that's happening now via @smyrick's PR [here](https://github.com/apollographql/federation/pull/2704).
